### PR TITLE
Man/interpolate

### PR DESCRIPTION
**Pull request by rouzbeh**

### CODE REVIEW

**Comment by joe-nguyen-actigraph** [comment]
@rouzbeh  Do you have any plots to show signal in and signal out pairs for the new resampling techniques?  The signal in can be something like in that Lund University paper.  Maybe a chirp signal swept from 1 Hz to fsamp*0.25 as input, then what the signal looks like after lowpass filtering and then the resampling.  I'm just curious how effective the lowpass is and then what distortion the mne resample does.  I'm kind of crunched for time so won't be able to install and run the code myself.  This garbage-in-garbage-out test will be the quickest way I can evaluate the effectiveness of the change.

**Comment by rouzbeh** [comment]
> @rouzbeh Do you have any plots to show signal in and signal out pairs for the new resampling techniques? The signal in can be something like in that Lund University paper. Maybe a chirp signal swept from 1 Hz to fsamp*0.25 as input, then what the signal looks like after lowpass filtering and then the resampling. I'm just curious how effective the lowpass is and then what distortion the mne resample does. I'm kind of crunched for time so won't be able to install and run the code myself. This garbage-in-garbage-out test will be the quickest way I can evaluate the effectiveness of the change.
legacy
new

legacy_after_bpf
raph/agcounts/assets/511375/595038b6-95f5-482a-861a-fbd0b8ebd307)

File: agcounts/legacy.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        epoch_counts[0, i] = np.floor(
+            sum(down_sample10_hz[0, i * block_size : i * block_size + block_size])
+        )
+    del down_sample10_hz
+    gc.collect(1)
+    return epoch_counts
+
+
+def _factors(frequency: int):
+    if frequency == 30:
```
Comment (dgerstel) [inline agcounts/legacy.py:213]: Can use a dictionary `frequency -> (upsample_factor, downsample_factor)` defaulting to `(1, 1)` e.g. `dictionary.get(default=(1, 1))`

File: tests/test_pow2.py [hunk 2]
```diff
@@ -25,5 +25,10 @@ def test_interpolated(taso_256_lpf, output_30hz):
...
 
 def test_32hz(input_32hz, output_30hz):
-    resampled = resample_to_30hz(input_32hz, 32)
+    resampled = resample_to_30hz(input_32hz, sample_rate=32)
     assert np.abs(resampled - output_30hz).max() == pytest.approx(0)
+
+
+def test_32hz_interpolate(input_32hz, output_30hz):
+    resampled = resample_to_30hz(input_32hz, sample_rate=32, interpolate=True)
+    assert np.abs(resampled - output_30hz).max() < 1
```
Comment (dgerstel) [inline tests/test_pow2.py:34]: We're allowing for max 1-g difference between expected and obtained?
Shall we test for mean, std, etc.?

File: agcounts/pow2.py [hunk 3]
```diff
@@ -1,14 +1,37 @@
...
+try:
+    from mne import filter
+except ImportError:
+    filter = None
+
+logger = logging.getLogger(__name__)
 
-def resample_to_30hz(data, sample_rate):
+
+def resample_to_30hz(data, sample_rate, interpolate: bool = False):
```
Comment (dgerstel) [inline agcounts/pow2.py:16]: Can we add here some docstring to explain the parameters?
It seems in both cases of if-else we interpolate (I guess the first kind is a direct interpolation and the other one is after upsampling to 256 Hz?)

File: tests/test_counts.py [hunk 4]
```diff
@@ -41,4 +41,16 @@ def test_counts_cp(agcounts, cp_counts):
...
+@pytest.mark.parametrize(
+    "agcounts_better_lpf,al_counts",
+    [
+        ((30, 40), (30, 40)),
+        ((10, 40), (10, 40)),
+    ],
+    indirect=True,
+)
+def test_counts_al_better_lpf(agcounts_better_lpf, al_counts):
+    assert agcounts_better_lpf.shape == al_counts.shape
```
Comment (dgerstel) [inline tests/test_counts.py:56]: Just for my understanding, the new implementation gives exactly the same counts as the old one?
(These tests show that, but just wanted be sure I didn't miss anything)

**Review by dgerstel — APPROVED** [review]
Some questions and optional suggestions

File: tests/test_counts.py (hunk 4 above)
Comment (rouzbeh) [reply tests/test_counts.py:56]: Only the shapes are identical. The counts change!

File: tests/test_pow2.py [hunk 5]
```diff
@@ -25,5 +25,11 @@ def test_interpolated(taso_256_lpf, output_30hz):
...
 def test_32hz(input_32hz, output_30hz):
-    resampled = resample_to_30hz(input_32hz, 32)
+    resampled = resample_to_30hz(input_32hz, sample_rate=32)
     assert np.abs(resampled - output_30hz).max() == pytest.approx(0)
+
+
+def test_32hz_interpolate(input_32hz, output_30hz):
+    resampled = resample_to_30hz(input_32hz, sample_rate=32, use_mne_filter=True)
+    assert np.abs(resampled - output_30hz).max() < 1
+    assert resampled.mean() == pytest.approx(-5.8447899368136745e-05)
```
Comment (dgerstel) [inline tests/test_pow2.py:35]: I don't want to be nitpicking, but the mean of the difference would be a stronger test
`np.abs(resampled - output_30hz).mean() < xxx`

**Pull request merged**